### ISSUE TITLE
Generated Solutions aren't unique

### ISSUE THREAD

**Issue by Jarno458** [MorganR/SudokuSpice#1]
Using the following code:
![debug1](https://user-images.githubusercontent.com/5691920/201425829-015311ef-8cc4-4f7b-9d5c-6ada5dfbc08e.JPG)
```
            var hintsCount = 21;

            var generator = new StandardPuzzleGenerator();
            var solver = StandardPuzzles.CreateSolver();
            PuzzleWithPossibleValues puzzle;

            do
            {
                puzzle = generator.Generate(9, hintsCount, TimeSpan.FromSeconds(5));
            } while (!solver.HasUniqueSolution(puzzle));

            var x = solver.HasUniqueSolution(puzzle);

            var stats = solver.ComputeStatsForAllSolutions(puzzle);

            LogWrite(puzzle.ToString(), Color.White);
```

I generated this puzzle:
```
╔═╤═╤═╦═╤═╤═╦═╤═╤═╗
║ │ │ ║ │ │ ║1│ │ ║
╟─┼─┼─╫─┼─┼─╫─┼─┼─╢
║ │ │ ║5│ │3║4│ │ ║
╟─┼─┼─╫─┼─┼─╫─┼─┼─╢
║ │9│3║ │ │ ║2│5│ ║
╠═╪═╪═╬═╪═╪═╬═╪═╪═╣
║ │ │ ║ │ │4║ │ │ ║
╟─┼─┼─╫─┼─┼─╫─┼─┼─╢
║ │3│ ║ │5│ ║ │ │ ║
╟─┼─┼─╫─┼─┼─╫─┼─┼─╢
║ │ │ ║ │ │9║ │ │ ║
╠═╪═╪═╬═╪═╪═╬═╪═╪═╣
║3│ │ ║9│ │2║8│ │ ║
╟─┼─┼─╫─┼─┼─╫─┼─┼─╢
║ │ │ ║ │ │ ║7│ │ ║
╟─┼─┼─╫─┼─┼─╫─┼─┼─╢
║ │ │ ║7│6│5║ │4│ ║
╚═╧═╧═╩═╧═╧═╩═╧═╧═╝
```

Dispite solver.HasUniqueSolution() returning true and solver.ComputeStatsForAllSolutions() saying it has a single solution it has infact many, below is a screenshot of an online sudoku solver https://sudokuspoiler.com/sudoku/sudoku9

![OnlineSudokuSolver](https://user-images.githubusercontent.com/5691920/201425801-eda96fb0-49bb-456a-8d59-0492b60ef050.JPG)

**Comment by MorganR** [comment]
I'm having trouble reproducing this issue. If I run this code, the asserts pass:

```C#
var solver = StandardPuzzles.CreateSolver();

int?[][]matrix = new int?[][] {
   new int?[] {null, null, null, null, null, null,    1, null, null},
   new int?[] {null, null, null,    5, null,    3,    4, null, null},
   new int?[] {null,    9,    3, null, null, null,    2,    5, null},
   new int?[] {null, null, null, null, null,    4, null, null, null},
   new int?[] {null,    3, null, null,    5, null, null, null, null},
   new int?[] {null, null, null, null, null,    9, null, null, null},
   new int?[] {   3, null, null,    9, null,    2,    8, null, null},
   new int?[] {null, null, null, null, null, null,    7, null, null},
   new int?[] {null, null, null,    7,    6,    5, null,    4, null},
};
var puzzle = new PuzzleWithPossibleValues(matrix);

Assert.False(solver.HasUniqueSolution(puzzle));
Assert.True(solver.ComputeStatsForAllSolutions(puzzle).NumSolutionsFound > 1);
```

Same for the constraint-based version.

Are you using the rule-based or constraint-based solver?

Are you using a custom implementation of `IPuzzle` or `IPuzzleWithPossibleValues`? If so, it's possible that this isn't playing well with the framework. Could you please point me to your implementation?

**Comment by MorganR** [comment]
Hmmm never mind, I see you're using the StandardPuzzleGenerator, so this is with rule-based puzzles. I was able to reproduce this using teh generator. 

It looks like there is potentially an issue with the `PossibleValues` tracking during puzzle generation, since, as demonstrated above, the solver can identify that the resulting puzzle is non-unique when its created from scratch, but not when using the output directly from the generator.

**Comment by Jarno458** [comment]
I am glad you can reproduce it, hope it's an easy fix

**Comment by MorganR** [comment]
This is fixed in release 3.2.2. Thank you for filing the issue!
